File: Rust/src/_01_algorithmic_toolbox/week2_algorithmic_warmup/math.rs
pub fn fibonacci(n: u64) -> u64 {
    let mut pair = (0, 1);
    for _ in 0..n {
        let (a, b) = pair;
        pair = (b, a + b);
    }
    pair.0
}

// return nth fibonacci modulo
pub fn fibonacci_modulo(n: u64, modulo: u64) -> u64 {
    let mut pair = (0, 1);
    for _ in 0..n {
        let (a, b) = pair;
        pair = (b, (a + b) % modulo);
    }
    pair.0
}

// last digit of nth fibonacci
pub fn fibonacci_last_digit(n: u64) -> u64 {
    fibonacci_modulo(n, 10)
}

// greatest common divisor
pub fn gcd(a: u64, b: u64) -> u64 {
    match b {
        0 => a,
        _ => gcd(b, a % b),
    }
}

// least common multiple
pub fn lcm(a: u64, b: u64) -> u64 {
    a * b / gcd(a, b)
}

// returns the period length of fibonacci modulo (Pisano period)
pub fn pisano_period(modulo: u64) -> u64 {
    if modulo == 1 {
        return 1;
    }
    let mut pair = (0, 1);
    let mut sequence_length = 0;
    loop {
        let (a, b) = pair;
        pair = (b, (a + b) % modulo);
        sequence_length += 1;
        if pair == (0, 1) {
            return sequence_length;
        };
    }
}

// returns the modulo of the nth fibonacci (faster version)
// uses fact that fibonacci modulo sequence repeats after
// length of pisano period
pub fn fibonacci_modulo_fast(n: u64, modulo: u64) -> u64 {
    let pisano_period = pisano_period(modulo);
    fibonacci_modulo(n % pisano_period, modulo)
}

// returns the last digit of the sum of fibonacci numbers up to n
pub fn fibonacci_sum_last_digit(n: u64) -> u64 {
    // can show that reduces to fib(n+2) - fib(2)
    (10 + (fibonacci_modulo_fast(n + 2, 10) - 1)) % 10
}

// returns the last digit of the sum of fibonacci numbers from n to m
pub fn fibonacci_partial_sum_last_digit(n: u64, m: u64) -> u64 {
    // can show that reduces to fib(m+2) - fib(n+1)
    (10 + fibonacci_modulo_fast(m + 2, 10) - fibonacci_modulo_fast(n + 1, 10)) % 10
}

// returns the last digit of sum of squares of fibonacci numbers up to n
pub fn fibonacci_sum_of_squares_last_digit(n: u64) -> u64 {
  // can show that reduces to fib(n) * fib(n+1)
  // from fact that fib(n) = fib(n+1) - fib(n-1)
  // fib(n)^2 = fib(n) * (fib(n+1) - fib(n-1))
  // fib(n)^2 = fib(n) * fib(n+1) - fib(n) * fib(n-1)
  // then write out as sequence
  fibonacci_modulo_fast(n, 10).pow(2) % 10
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_correct_fibonacci() {
        assert_eq!(fibonacci(50), 12586269025);
    }

    #[test]
    fn test_correct_fibonacci_last_digit() {
        assert_eq!(fibonacci_last_digit(10), 5)
    }

    #[test]
    fn test_correct_gcd() {
        assert_eq!(gcd(28851538, 1183019), 17657)
    }

    #[test]
    fn test_correct_lcm() {
        assert_eq!(lcm(28851538, 1183019), 1933053046)
    }

    #[test]
    fn test_correct_pisano_period() {
        assert_eq!(pisano_period(10), 60)
    }

    #[test]
    fn test_correct_fibonacci_modulo_fast() {
        assert_eq!(fibonacci_modulo_fast(2816213588, 239), 151)
    }

    #[test]
    fn test_correct_fibonacci_sum_last_digit() {
        assert_eq!(fibonacci_sum_last_digit(100), 5)
    }

    #[test]
    fn test_correct_fibonacci_partial_sum_last_digit() {
        assert_eq!(fibonacci_partial_sum_last_digit(10, 200), 2)
    }

    #[test]
    fn test_correct_fibonacci_sum_of_squares_last_digit() {
        assert_eq!(fibonacci_sum_of_squares_last_digit(1234567890), 0)
    }
}

File: Typescript/src/01.AlgorithmicToolbox/Week2AlgorithmicWarmup/math.ts
export const fibonacci = (n: number): number => {
  let pair: [number, number] = [0, 1];
  for (let i = 0; i < n; ++i) {
    const [a, b] = pair;
    pair = [b, a + b];
  }
  return pair[0];
};

// returns nth fibonacci modulo
const fibonacciModulo = (n: number, modulo: number): number => {
  let pair: [number, number] = [0, 1];
  for (let i = 0; i < n; ++i) {
    const [a, b] = pair;
    pair = [b, (a + b) % modulo];
  }
  return pair[0];
};

// returns last digit of fibonacci
export const fibonacciLastDigit = (n: number): number => {
  return fibonacciModulo(n, 10);
};

// greatest common divisor
export const gcd = (a: number, b: number): number => {
  if (b === 0) {
    return a;
  }
  return gcd(b, a % b);
};

// least common multiple
export const lcm = (a: number, b: number): number => {
  return (a * b) / gcd(a, b);
};

// returns the period length of the fibonacci modulo
export const pisanoPeriod = (modulo: number): number => {
  if (modulo === 1) {
    return 1;
  }
  let pair: [number, number] = [0, 1];
  let sequenceLength = 0;
  while (true) {
    const [a, b] = pair;
    pair = [b, (a + b) % modulo];
    sequenceLength += 1;
    if (pair[0] === 0 && pair[1] === 1) {
      // sequence has repeated
      return sequenceLength;
    }
  }
};

// returns the modulo of the nth fibonacci (faster version)
// uses fact that fibonacci modulo sequence repeats after
// length of pisano period
export const fibonacciModuloFast = (n: number, modulo: number): number => {
  return fibonacciModulo(n % pisanoPeriod(modulo), modulo);
};

// returns the last digit of the sum of fibonacci numbers
export const fibonacciSumLastDigit = (n: number): number => {
  // can show that reduces to fib(n+2) - fib(2)
  return (10 + fibonacciModuloFast(n + 2, 10) - 1) % 10;
};

// returns the last digit of the sum of fibonacci numbers from n to m
export const fibonacciPartialSumLastDigit = (n: number, m: number): number => {
  if (m < n) {
    throw Error("m cannot be less than n");
  }
  // can show that reduces to fib(m+2) - fib(n+1)
  return (
    (10 + fibonacciModuloFast(m + 2, 10) - fibonacciModuloFast(n + 1, 10)) % 10
  );
};

// returns the last digit of sum of squares of fibonacci numbers
export const fibonacciSumOfSquaresLastDigit = (n: number): number => {
  // can show that reduces to fib(n) * fib(n+1)
  // from fact that fib(n) = fib(n+1) - fib(n-1)
  // fib(n)^2 = fib(n) * (fib(n+1) - fib(n-1))
  // fib(n)^2 = fib(n) * fib(n+1) - fib(n) * fib(n-1)
  // then write out as sequence
  return (fibonacciModuloFast(n, 10) * fibonacciModuloFast(n + 1, 10)) % 10;
};
